test/e2e/storage/csi_volumes.go

```diff
@@ -0,0 +1,544 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storage
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"strings"
+	"time"
+
+	"k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	storagev1alpha1 "k8s.io/api/storage/v1alpha1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+const (
+	csiNodeAnnotationMappingKey string = "csi.volume.kubernetes.io/nodeid"
+	csiMockPluginImage          string = "docker.io/k8scsi/mock-plugin:0.1"
+	csiExternalAttacher         string = "docker.io/k8scsi/csi-attacher:0.1"
+	csiSocketVolume             string = "/var/lib/csi/sockets/pluginproxy"
+	csiSocketPath               string = csiSocketVolume + "/mock.socket"
+	csiPvName                   string = "pv-csi-e2e-001"
+)
+
+func externalAttacherServiceAccout(
+	client clientset.Interface,
+	config framework.VolumeTestConfig,
+	teardown bool,
+) *v1.ServiceAccount {
+	serviceAccountName := config.Prefix + "-service-account"
+	serviceAccountClient := client.CoreV1().ServiceAccounts(config.Namespace)
+	sa := &v1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: serviceAccountName,
+		},
+	}
+
+	serviceAccountClient.Delete(sa.GetName(), &metav1.DeleteOptions{})
+	err := wait.Poll(2*time.Second, 10*time.Minute, func() (bool, error) {
+		_, err := serviceAccountClient.Get(sa.GetName(), metav1.GetOptions{})
+		return apierrs.IsNotFound(err), nil
+	})
+	framework.ExpectNoError(err, "Timed out waiting for deletion: %v", err)
+
+	if teardown {
+		return nil
+	}
+
+	ret, err := serviceAccountClient.Create(sa)
+	if err != nil {
+		framework.ExpectNoError(err, "Failed to create %s service account: %v", sa.GetName(), err)
+	}
+
+	return ret
+}
+
+func externalAttacherClusterRole(
+	client clientset.Interface,
+	config framework.VolumeTestConfig,
+	teardown bool,
+) *rbacv1.ClusterRole {
+	clusterRoleClient := client.RbacV1().ClusterRoles()
+	role := &rbacv1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "external-attacher-runner",
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups: []string{""},
+				Resources: []string{"persistentvolumes"},
+				Verbs:     []string{"get", "list", "watch", "update"},
+			},
+			{
+				APIGroups: []string{""},
+				Resources: []string{"nodes"},
+				Verbs:     []string{"get", "list", "watch"},
+			},
+			{
+				APIGroups: []string{"storage.k8s.io"},
+				Resources: []string{"volumeattachments"},
+				Verbs:     []string{"get", "list", "watch", "update"},
```

Comment:
Resource "nodes" need Verbs "update" from Line 423;
Resource "volumeattachments" need Verbs "delete" from Line 351.

Reply:
This is policy for external-attacher started as a pod somewhere in this PR. It's not policy for the test itself.

Reply:
@jsafrane Got it.